ledger/src/blockstore.rs

```diff
@@ -3345,29 +3345,61 @@ fn update_completed_data_indexes(
     let start_shred_index = completed_data_indexes
         .range(..new_shred_index)
         .next_back()
+        // `index` here is the last shred in the previous data set, so we need `+ 1`, as we are
+        // looking for the first shred in the current data set.
         .map(|index| index + 1)
-        .unwrap_or_default();
-    // Consecutive entries i, k, j in this vector represent potential ranges [i, k),
-    // [k, j) that could be completed data ranges
-    let mut shred_indices = vec![start_shred_index];
-    // `new_shred_index` is data complete, so need to insert here into the
-    // `completed_data_indexes`
+        .unwrap_or(0);
+
     if is_last_in_data {
         completed_data_indexes.insert(new_shred_index);
-        shred_indices.push(new_shred_index + 1);
     }
-    if let Some(index) = completed_data_indexes.range(new_shred_index + 1..).next() {
-        shred_indices.push(index + 1);
+
+    let end_shred_index = completed_data_indexes
+        .range(new_shred_index + 1..)
+        .next()
+        .copied();
+
+    // Checks that all shards in the `[begin, end]` range are present in `received_data_shreds`.
+    let is_complete_set = |begin: u32, end: u32| {
+        let num_shreds = usize::try_from(end - begin + 1).unwrap();
+        received_data_shreds
+            .range(u64::from(begin)..=u64::from(end))
+            .count()
+            == num_shreds
+    };
+
+    // There are several possibilities here.  Newly added shred could be the last one in a data set.
+    // It may also end a data set, creating two completed data sets at once.
+    //
+    // And it is possible that we do not know an index of the last shred in a subsequent data set
+    // (`end_shred_index`).
+    match (is_last_in_data, end_shred_index) {
+        (false, None) => vec![],
+        (false, Some(end_shred_index)) => {
+            if is_complete_set(start_shred_index, end_shred_index) {
+                vec![(start_shred_index, end_shred_index)]
+            } else {
+                vec![]
+            }
+        }
+        (true, None) => {
+            if is_complete_set(start_shred_index, new_shred_index) {
+                vec![(start_shred_index, new_shred_index)]
+            } else {
+                vec![]
+            }
+        }
+        (true, Some(end_shred_index)) => {
+            let mut res = vec![];
+            if is_complete_set(start_shred_index, new_shred_index) {
+                res.push((start_shred_index, new_shred_index));
+            }
+            if is_complete_set(new_shred_index + 1, end_shred_index) {
+                res.push((new_shred_index + 1, end_shred_index));
+            }
+            res
+        }
```

Comment:
Comparing the two sides, I kind of feel like the old code was actually more expressive and perhaps better.
Like I see 7 different branches here (which imo makes it harder to read and make it easier to introduce bugs), but the old code was avoiding too many branches. Besides all the repetitive `if is_complete_set` lines aren't ideal.

Maybe instead of this keep the old implementation but add more comments/context to the code where needed.

Reply:
Sorry for a long response.  Not sure how to make it shorter.

---

As an author of the new version, I am obviously biased.  But I am also the
reader of the old version, and it was pretty hard for me to understand.  It
required multiple iterations.  And only after I've rewritten it myself, I was
able to follow all the details.  I think, the new version is simpler in this
regard, but, maybe there is a better way to write it.  Here are the difficulties
I've faced with the older version:

1. Data is transformed with the `shred_indices` first being populated, and then
   consumed.  So I have to evaluate the code in my head to understand what is
   going on.

2. `shred_indices` as a name is not very descriptive, and it contributes to the
   mental load of the reader.

   I had to guess what kind of indices are stored in there.  The `+ 1` in index
   manipulations only makes it harder.  Now, I understand that it could be named
   something like `indices of the first shred in a data set`.  But I was not
   able to infer this until recently.

   While reading the code, I just had to keep it as a variable and hope to infer
   the behavior from the rest of the code.

3. `update_completed_data_indexes` returns ranges of shred indices that describe
   zero, one or two complete data sets.  In the returned value, these ranges are
   inclusive.  A return value of `(0, 1)` is a range that covers the first two
   shreds.

   But in this code:

```rust
    shred_indices
        .windows(2)
        .filter(|ix| {
            let (begin, end) = (ix[0] as u64, ix[1] as u64);
            let num_shreds = (end - begin) as usize;
            received_data_shreds.range(begin..end).count() == num_shreds
        })
        .map(|ix| (ix[0], ix[1] - 1))
        .collect()
```

   `begin` and `end` is an exclusive range of `[begin..end)`.  Which is also the
   reason for the `map(|ix| (ix[0], ix[1] - 1))` at the end.

   This inconsistency is required in order to be able to treat an index as both
   the end of one set and a beginning of the next.  Otherwise, `windows(2)`
   would not be able to generate correct slices.

In the new version, I was trying to address all of the above.

It is indeed longer, as I have unrolled the transformations.  But, I think, for
the reader, it is now easier to follow.  There is no state to evaluate in one's
head.  You just read it from the top to the bottom, and it covers all the cases.
And most of the cases are trivial.  The only part that is at least a bit complex
is the case when the new shard creates two ranges.  But it is a more complex
case by its nature, so it seems reasonable that the code would be a bit more
complex as well.

The new version also has less `+ 1` and no `- 1` index manipulations.  Which, I
think, both makes it more readable and reduces the chances that it can be
incorrect, as there are just fewer places to check.  All index ranges are
inclusive.

Not sure what do you mean by saying that the `if is_complete_set()` lines aren't
ideal.  I agree they are a bit verbose, but they are super simple, covering all
4 possible combinations of the range ends.

So, I would like to disagree with you on this statement: because the new version
is longer, it is more likely to have bugs.  Complex code is harder to read, and
harder to reason about.  So even being shorter, it may still contain bugs and
requires a more thorough checking.

Simpler is better :)  More so for readability, but also for correctness.  To a
certain degree, of course.

If you insist, I can try to document the old version and, possibly, rename some
of the variables.  But, as the old version uses a more general approach, it
fundamentally is more complex.  As currently we only insert one shred, this
generalization will always make the older version more complex that it needs to
be.

Reply:
Okay, it is fair to have different viewpoints on which version is more expressive/cleaner.

However the old code is still technically correct and working fine; and the implementation is well encapsulated inside a single function. So it is not urgent or high priority to change it right now.
Given above and the fact that there is a disagreement on which version is better, I would suggest lets not change this for now (except for maybe adding more comments/context where needed) in favor of prioritizing more urgent or necessary changes.